cni-plugin/pkg/install/install.go

```diff
@@ -496,3 +504,46 @@ func setSuidBit(file string) error {
 
 	return nil
 }
+
+func createCNIToken(kubecfg *rest.Config) string {
```

Comment:
Looks like we're duplicating very similar logic here and within the node directory. 

Probably would be better to do this in once place - we can use a single constant for the validity duration, as well as a single function for refreshing the token.

I'd recommend creating a new structure to handle this logic, as there is a little bit of complexity that we need to deal with. Specifically, the returned expiry time may not be the time that we requested, so we need to cache that and update our ticker based on the value that's returned (not just the value we requested).

> ExpirationSeconds is the requested duration of validity of the request. The token issuer may return a token with a different validity duration so a client needs to check the 'expiration' field in a response.

From here: https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-request-v1/

e.g., something like this:

```go

type TokenRefresher struct {
  cfg *rest.Config
  clientset *kubernetes.ClientSet
  ...
}

type TokenUpdate struct {
  Token string
  Error error
}

func (t *TokenRefresher) TokenChan() chan TokenUpdate {}
func (t* TokenRefresher) UpdateToken() TokenUpdate {}
```

The code in this file (install.go) can just call TokenUpdate() to get a new token. 

The code in node that updates the token can do something like this:

```go
tokenChan := t.TokenChan()
for {
  select {
    case <- tokenChan:
      // Update CNI config with new token.
  }
}
```

Reply:
Another thing to think about.. would be whether or not the `install.go` code should bother generating a token at all - we need to have the node code running, and updating tokens anyway, and it needs to do it ASAP because otherwise it won't know what the returned expiry time is on token generated by the CNI plugin installer.

I'll need to think about this one a bit more - LMK what you think.

Reply:
Looking at the actual validity period is a good point. I will adapt the code accordingly. When should we try to replace the token? In the initial change I just statically assumed 1/3 of the (expected) validity period so that we still have enough time for another attempt if the token request fails.
Concerning the code duplication, I completely agree, but did not want to touch/change too many things at once. Writing the kubeconfig is anyway completely duplicated in slightly different form. It would for sure be easier to have the code only in `calico-node`, but I understand it originated in the cni installation and only moved to `calico-node` when token got restricted validity.
In case you want to continue to have the kubeconfig writing logic in two places, where should I put the token update code?

Reply:
> When should we try to replace the token? In the initial change I just statically assumed 1/3 of the (expected) validity period so that we still have enough time for another attempt if the token request fails.

I think 1/3 of the period is a fine enough way to do it - I might even do a token refresh every hour for the base case, and then switch to doing 1/3 of the period if the returned expiry is less than an hour long.

That actually makes me realize that these tokens are valid cluster-wide, not within the context of a particular node, so putting the logic to refresh the token within `calico/node` might not be right - it means we might get 5000 nodes all trying to refresh their token at the same time.

Think we might need a controller in kube-controllers which refreshes the token, and then watchers within calico/node that spot when the token has changed and update the config for that node. 

> Writing the kubeconfig is anyway completely duplicated in slightly different form

Yep, longer term I'd like to consolidate this as well. It's remnant of cni-plugin and node being separate repositories in the past. I'd like to see us consolidate all of this into one place or the other eventually. 

> In case you want to continue to have the kubeconfig writing logic in two places, where should I put the token update code?

If the above is true, I think we probably want the token refreshing code to live in kube-controllers, and then all calico/node and the CNI plugin code needs to do is query the generated token via a Get()

Reply:
I think the main argument against having the token code in `kube-controllers` is that `kube-controllers` is running in the pod network as far as I know. Therefore, it needs a running CNI to setup its network sandbox. Hence, it cannot be the precondition for CNI. Not sure how to follow up here.

Reply:
> I think the main argument against having the token code in kube-controllers is that kube-controllers is running in the pod network as far as I know.

Yeah, good call. This would be a problem. I'll think on this some more.

One option would be to start running kube-controllers as host networked, but I'd really prefer not to do that.